File: slidr/src/main/java/com/r0adkll/slidr/widget/ScrimRenderer.java
package com.r0adkll.slidr.widget;


import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Rect;
import android.view.View;

import androidx.annotation.NonNull;

import com.r0adkll.slidr.model.SlidrPosition;


final class ScrimRenderer {

    private final View rootView;
    private final View decorView;
    private final Rect dirtyRect;

    ScrimRenderer(@NonNull View rootView, @NonNull View decorView) {
        this.rootView = rootView;
        this.decorView = decorView;
        dirtyRect = new Rect();
    }

    void render(Canvas canvas, SlidrPosition position, Paint paint) {
        switch (position) {
            case Left:
                renderLeft(canvas, paint);
                break;
            case Right:
                renderRight(canvas, paint);
                break;
            case Top:
                renderTop(canvas, paint);
                break;
            case Bottom:
                renderBottom(canvas, paint);
                break;
            case Vertical:
                renderVertical(canvas, paint);
                break;
            case Horizontal:
                renderHorizontal(canvas, paint);
                break;
        }
    }


    Rect getDirtyRect(SlidrPosition position) {
        switch (position) {
            case Left:
                dirtyRect.set(0, 0, decorView.getLeft(), rootView.getMeasuredHeight());
                break;
            case Right:
                dirtyRect.set(decorView.getRight(), 0, rootView.getMeasuredWidth(), rootView.getMeasuredHeight());
                break;
            case Top:
                dirtyRect.set(0, 0, rootView.getMeasuredWidth(), decorView.getTop());
                break;
            case Bottom:
                dirtyRect.set(0, decorView.getBottom(), rootView.getMeasuredWidth(), rootView.getMeasuredHeight());
                break;
            case Vertical:
                if (decorView.getTop() > 0) {
                    dirtyRect.set(0, 0, rootView.getMeasuredWidth(), decorView.getTop());
                } else {
                    dirtyRect.set(0, decorView.getBottom(), rootView.getMeasuredWidth(), rootView.getMeasuredHeight());
                }
                break;
            case Horizontal:
                if (decorView.getLeft() > 0) {
                    dirtyRect.set(0, 0, decorView.getLeft(), rootView.getMeasuredHeight());
                } else {
                    dirtyRect.set(decorView.getRight(), 0, rootView.getMeasuredWidth(), rootView.getMeasuredHeight());
                }
                break;
        }
        return dirtyRect;
    }


    private void renderLeft(Canvas canvas, Paint paint) {
        canvas.drawRect(0, 0, decorView.getLeft(), rootView.getMeasuredHeight(), paint);
    }


    private void renderRight(Canvas canvas, Paint paint) {
        canvas.drawRect(decorView.getRight(), 0, rootView.getMeasuredWidth(), rootView.getMeasuredHeight(), paint);
    }


    private void renderTop(Canvas canvas, Paint paint) {
        canvas.drawRect(0, 0, rootView.getMeasuredWidth(), decorView.getTop(), paint);
    }


    private void renderBottom(Canvas canvas, Paint paint) {
        canvas.drawRect(0, decorView.getBottom(), rootView.getMeasuredWidth(), rootView.getMeasuredHeight(), paint);
    }


    private void renderVertical(Canvas canvas, Paint paint) {
        if (decorView.getTop() > 0) {
            renderTop(canvas, paint);
        } else {
            renderBottom(canvas, paint);
        }
    }


    private void renderHorizontal(Canvas canvas, Paint paint) {
        if (decorView.getLeft() > 0) {
            renderLeft(canvas, paint);
        } else {
            renderRight(canvas, paint);
        }
    }
}

File: slidr/src/main/kotlin/com/r0adkll/slidr/widget/ScrimRenderer.kt
package com.r0adkll.slidr.widget

import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Rect
import android.view.View
import com.r0adkll.slidr.model.SlidrPosition

internal class ScrimRenderer(private val rootView: View, private val decorView: View) {

    private val dirtyRect = Rect()

    fun render(canvas: Canvas, position: SlidrPosition?, paint: Paint) {
        when (position) {
            SlidrPosition.Left -> renderLeft(canvas, paint)
            SlidrPosition.Right -> renderRight(canvas, paint)
            SlidrPosition.Top -> renderTop(canvas, paint)
            SlidrPosition.Bottom -> renderBottom(canvas, paint)
            SlidrPosition.Vertical -> renderVertical(canvas, paint)
            SlidrPosition.Horizontal -> renderHorizontal(canvas, paint)
        }
    }

    fun getDirtyRect(position: SlidrPosition?): Rect {
        when (position) {
            SlidrPosition.Left -> dirtyRect[0, 0, decorView.left] = rootView.measuredHeight

            SlidrPosition.Right -> dirtyRect[decorView.right, 0, rootView.measuredWidth] =
                rootView.measuredHeight

            SlidrPosition.Top -> dirtyRect[0, 0, rootView.measuredWidth] = decorView.top

            SlidrPosition.Bottom -> dirtyRect[0, decorView.bottom, rootView.measuredWidth] =
                rootView.measuredHeight

            SlidrPosition.Vertical -> if (decorView.top > 0) {
                dirtyRect[0, 0, rootView.measuredWidth] = decorView.top
            } else {
                dirtyRect[0, decorView.bottom, rootView.measuredWidth] = rootView.measuredHeight
            }

            SlidrPosition.Horizontal -> if (decorView.left > 0) {
                dirtyRect[0, 0, decorView.left] = rootView.measuredHeight
            } else {
                dirtyRect[decorView.right, 0, rootView.measuredWidth] = rootView.measuredHeight
            }
        }
        return dirtyRect
    }

    private fun renderLeft(canvas: Canvas, paint: Paint) {
        canvas.drawRect(0f, 0f, decorView.left.toFloat(), rootView.measuredHeight.toFloat(), paint)
    }

    private fun renderRight(canvas: Canvas, paint: Paint) {
        canvas.drawRect(
            decorView.right.toFloat(),
            0f,
            rootView.measuredWidth.toFloat(),
            rootView.measuredHeight.toFloat(),
            paint,
        )
    }

    private fun renderTop(canvas: Canvas, paint: Paint) {
        canvas.drawRect(0f, 0f, rootView.measuredWidth.toFloat(), decorView.top.toFloat(), paint)
    }

    private fun renderBottom(canvas: Canvas, paint: Paint) {
        canvas.drawRect(
            0f,
            decorView.bottom.toFloat(),
            rootView.measuredWidth.toFloat(),
            rootView.measuredHeight.toFloat(),
            paint,
        )
    }

    private fun renderVertical(canvas: Canvas, paint: Paint) {
        if (decorView.top > 0) {
            renderTop(canvas, paint)
        } else {
            renderBottom(canvas, paint)
        }
    }

    private fun renderHorizontal(canvas: Canvas, paint: Paint) {
        if (decorView.left > 0) {
            renderLeft(canvas, paint)
        } else {
            renderRight(canvas, paint)
        }
    }
}
